File: c039(the_three_n_plus_one_problem)/Python/the_three_n_plus_one_problem.py
from sys import stdin, stdout

def func(n: int) -> int:
    count = 1
    while n != 1:
        if n % 2: n = 3 * n + 1
        else: n //= 2
        count += 1
    return count

record = {}

while True:
    line = stdin.readline().rstrip('\n')
    if not line: break
    stdout.write(line + ' ')
    i, j = map(int, line.split())
    if i > j: i, j = j, i
    max = 0
    for n in range(i, j + 1):
        if n in record: curr = record[n]
        else:
            curr = func(n)
            record[n] = curr
        if curr > max: max = curr
    stdout.write(str(max) + '\n')

File: c039(the_three_n_plus_one_problem)/C++/the_three_n_plus_one_problem.cpp
#include <iostream>
#include <map>
#include <algorithm>

int func(int n) {
  int count = 1;
  while (n != 1) n = (n % 2? 3 * n + 1: n / 2), count++;
  return count;
}

std::map<int, int> record;

int main(void) {
  std::ios::sync_with_stdio(false);
  std::cin.tie(0);
  int i, j;
  while (std::cin >> i >> j) {
    std::cout << i << ' ' << j << ' ';
    if (i > j) std::swap(i, j);
    int max = 0, curr;
    for (;i <= j; i++) {
      if (record.find(i) != record.end()) curr = record[i];
      else curr = func(i), record[i] = curr;
      if (curr > max) max = curr;
    }
    std::cout << max << '\n';
  }
  return 0;
}
